### PR TITLE
 implement jwt authentication 

### PR DESCRIPTION
**Pull request by Rodgers-M**
**What does this PR do?**
Implement token based authentication using [JWT](https://jwt.io/)

**Description of tasks to be completed?**
- create auth blueprint
- write tests for register and login 
- implement user registration
- implement user login and issuing access token

**How should this be tested manually?**
run `python manage.py test` form the application root directory
run the application and test the `register` and `login` routes using  `postman`

**Relevant tracker stories**

- 152997714

### CODE REVIEW

File: app/models.py [hunk 1]
```diff
@@ -31,6 +32,46 @@ def password(self, password):
 	def verify_password(self, password):
 		return check_password_hash(self.password_hash, password)
 
+	def save(self):
+		"""add user instance to session and save to databas"""
+		db.session.add(self)
+		db.session.commit()
+
+	def generate_token(self, user_id):
```
Comment (meshnesh) [inline app/models.py]: good implementation of the generate token

File: app/models.py [hunk 2]
```diff
@@ -31,6 +32,46 @@ def password(self, password):
...
+			}
+			# create the byte string token using the payload and the SECRET key
+			jwt_string = jwt.encode(
+				payload,
+				current_app.config.get('SECRET_KEY'),
+				algorithm='HS256'
+			)
+			return jwt_string
+
+		except Exception as e:
```
Comment (meshnesh) [inline app/models.py:58]: use pylint to show you best practices for naming conventions, like `e` should raise a concern

File: app/models.py [hunk 3]
```diff
@@ -31,6 +32,46 @@ def password(self, password):
...
+			# return an error in string format if an exception occurs
+			return str(e)
+
+	@staticmethod
+	def decode_token(token):
```
Comment (meshnesh) [inline app/models.py:63]: good implementation of decode token

File: manage.py [hunk 4]
```diff
@@ -14,5 +15,15 @@ def make_shell_context():
 manager.add_command("shell", Shell(make_context=make_shell_context))
 manager.add_command('db', MigrateCommand)
 
+#define a command for running tests
```
Comment (meshnesh) [inline manage.py:18]: good implementation. Does it work when you run it?

File: tests/test_auth.py [hunk 5]
```diff
@@ -0,0 +1,71 @@
...
+		"""Test app cannot registered same user twice."""
+		res = self.client().post('/auth/register', data=self.user_data, content_type='application/json')
+		self.assertEqual(res.status_code, 201)
+		second_res = self.client().post('/auth/register', data=self.user_data, content_type='application/json')
+		self.assertEqual(second_res.status_code, 202)
+		# get the results returned in json format
+		result = json.loads(second_res.data.decode())
+		self.assertEqual( result['message'], "a user with given email exists, please login")
+
+	def test_existing_user_login(self):
```
Comment (meshnesh) [inline tests/test_auth.py:48]: good testing for testing the token

**Comment by meshnesh** [comment]
Awesome Implementing JWT and restructuring your files.
Install pylint to help you in finding minor issues

File: manage.py (hunk 4 above)
Comment (Rodgers-M) [reply manage.py:18]: yes it works

File: app/models.py (hunk 3 above)
Comment (betsybeth) [reply app/models.py:63]: good commenting of code

**Comment by betsybeth** [comment]
Good work

**Pull request merged**